### DISCUSSION TITLE
How `log2_generic` get the `source_id` value ?

### DISCUSSION THREAD

**Discussion by shahawi-sumup** [zephyrproject-rtos/zephyr#50664] [Q&A] [answered]
I'm working on integrating our logging module with Zephyr logger v2 subsystem, this is working fine with logger v1 but not with v2.
Unfortunately, couldn't reproduce on another platforms like `native_posix` or `stm32f429i_disc1`, our platform is based on Cortex-A7 and the arch support code is made by the Chip vendor.

So the problem i'm facing is submitting the msgs via `log2_generic` call, but the messages got `source_id = -1`  which results in `ASSERTION FAIL [source_id < log_sources_count()] @ WEST_TOPDIR/os/subsys/logging/log_mgmt.c:174`.

My question is How `log2_generic` acquire the value of `source_id` ?

Zephyr v2.7 (LTS2)
Sample code
```
#include <logging/log.h>
#include <zephyr.h>

LOG_MODULE_REGISTER(ext_log_system);


void custom_logger(int levelconst char* formatStr, ...)
{
    LOG_INF("Sample log msg");      // Working
    va_list args;
    va_start(args, formatStr);

    log2_generic(3, formatStr, args);  //Not working
    va_end(args);
}

```

**Comment by shahawi-sumup** [comment]
CC: @nordic-krch if you can help, Thank you .

**Comment by yashi** [comment]
Is `log_source_id_get()` returning the value intended?
```
int log_source_id_get(const char *name)
{
	for (int i = 0; i < log_src_cnt_get(CONFIG_LOG_DOMAIN_ID); i++) {
		if (strcmp(log_source_name_get(CONFIG_LOG_DOMAIN_ID, i),
			   name) == 0) {
			return i;
		}
	}
	return -1;
}
```

**Comment by nordic-krch** [comment, marked as answer]
@shahawi-sumup you are right, `log2_generic` is missing source_id. I will try to fix it. Meanwhile you can use directly what is called underneath however it is a bit tricky. Source is a `void *` which points to different thing depending on `CONFIG_LOG_RUNTIME_FILTERING` so something like this can be used:

```
void *source = IS_ENABLED(CONFIG_LOG_RUNTIME_FILTERING) ? (void *)__log_current_dynamic_data : (void *)__log_current_const_data;
z_log_msg_runtime_vcreate(CONFIG_LOG_DOMAIN_ID, source, level, NULL, 0, 0, fmt,ap);`
```

`__log_current_dynamic_data`  and `__log_current_const_data` are declared by `LOG_MODULE_REGISTER()`

**Reply by edmondg-xlnx** [reply]
I don't think this has been fixed in either v3.5.0 nor v4.0.1

**Comment by shahawi-sumup** [comment]
@nordic-krch Thanks for your help, it worked!

I have added the following ASSERT to identify a problamtic msg @ line https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.7.1/subsys/logging/log_core.c#L706
```
__ASSERT(source_id < log_sources_count(), "Invalid source_id %d msg: %s", source_id, msg.msg->str);
```
because this ASSERT wasn't helping https://github.com/zephyrproject-rtos/zephyr/blob/zephyr-v2.7.1/subsys/logging/log_mgmt.c#L67

Could you please include something similar so it gets easier to identify the message behind the assert ?

---
Now I'm trying to find out why i get a lot of msgs dropped with logger v2 (but not with logger v1)
```
# These settings didn't affect the number of dropped msgs
CONFIG_LOG_PROCESS_THREAD_STACK_SIZE=2048
CONFIG_LOG_BUFFER_SIZE=20480
```

This makes it work, but affects system performance
```
CONFIG_LOG_BLOCK_IN_THREAD=y
```

Thanks alot
